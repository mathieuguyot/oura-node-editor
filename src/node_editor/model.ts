export interface XYPosition {
    x: number;
    y: number;
}

export interface LinkModel {
    inputNodeId: number;
    inputPinId: number;
    ouputNodeId: number;
    outputPinId: number;
    inputPinPosition ?: XYPosition;
    outputPinPosition ?: XYPosition;
}

export interface NodeModel {
    nId: number;
    name: string;
    x: number;
    y: number;
    width: number;
    connectors: ConnectorModel[];
}

export enum ConnectorType {
    Input,
    Output
}

export interface ConnectorModel {
    id: number;
    name: string;
    connectorType: ConnectorType;
    contentType?: string;
}
