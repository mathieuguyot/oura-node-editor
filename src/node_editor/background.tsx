import React from "react";
import { PanZoomModel } from "./model";

export interface BackGroundProps {
    panZoomInfo: PanZoomModel;
}

export default function BackGround(props: React.PropsWithChildren<BackGroundProps>): JSX.Element {
    const { panZoomInfo, children } = props;
    const grid = String(200 * panZoomInfo.zoom);
    return (
        <div
            style={{
                position: "relative",
                top: "0",
                left: "0",
                width: "100%",
                height: "100%",
                overflow: "hidden",
                backgroundColor: "#232323",
                backgroundPosition: `${panZoomInfo.topLeftCorner.x}px ${panZoomInfo.topLeftCorner.y}px`,
                backgroundImage: `url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='${grid}' height='${grid}' viewBox='0 0 100 100'%3E%3Cg fill-rule='evenodd'%3E%3Cg fill='black' fill-opacity='0.4'%3E%3Cpath opacity='.5' d='M96 95h4v1h-4v4h-1v-4h-9v4h-1v-4h-9v4h-1v-4h-9v4h-1v-4h-9v4h-1v-4h-9v4h-1v-4h-9v4h-1v-4h-9v4h-1v-4h-9v4h-1v-4H0v-1h15v-9H0v-1h15v-9H0v-1h15v-9H0v-1h15v-9H0v-1h15v-9H0v-1h15v-9H0v-1h15v-9H0v-1h15v-9H0v-1h15V0h1v15h9V0h1v15h9V0h1v15h9V0h1v15h9V0h1v15h9V0h1v15h9V0h1v15h9V0h1v15h9V0h1v15h4v1h-4v9h4v1h-4v9h4v1h-4v9h4v1h-4v9h4v1h-4v9h4v1h-4v9h4v1h-4v9h4v1h-4v9zm-1 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-9-10h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm9-10v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-9-10h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm9-10v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-9-10h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm9-10v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-9-10h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9z'/%3E%3Cpath d='M6 5V0H5v5H0v1h5v94h1V6h94V5H6z'/%3E%3C/g%3E%3C/g%3E%3C/svg%3E")`
            }}>
            {children}
        </div>
    );
}
