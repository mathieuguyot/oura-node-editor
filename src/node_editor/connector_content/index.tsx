import React from 'react';
import { ConnectorType } from '../model';
import { ConnectorContentProps } from './common';
import StringConnectorContent from './string';

export function createConnectorComponent(props: ConnectorContentProps)
{
    if(props.connectorModel.contentType === "string")
    {
        return <StringConnectorContent {...props}/>;
    }
    // Defaut return connector name
    const ct = props.connectorModel.connectorType;
    return <div style={{textAlign: ct === ConnectorType.Output ? "right" : "left"}}>
        {props.connectorModel.name}
    </div>;
}