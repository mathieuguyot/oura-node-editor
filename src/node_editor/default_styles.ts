import CSS from "csstype";

export const default_styles : {[nId: string] : {[nId: string] : CSS.Properties;};} = {
    dark: {
        node_selected: {
            boxShadow: "0px 0px 0px 2px white",
            borderRadius: "10px 10px 10px 10px",
        },
        node_unselected: {},
        node_header: {
            backgroundColor: "rgba(48, 141, 97, 0.75)",
            color: "white",
            borderRadius: "10px 10px 0px 0px",
            height: "20px",
            cursor: "grab",
        },
        node_background: {
            color: "white",
            backgroundColor: "rgba(63, 63, 63, 0.75)",
            cursor: "grab",
            userSelect: "none",
            MozUserSelect: "none",
        },
        node_footer: {
            borderRadius: "0px 0px 10px 10px",
            height: "10px",
            backgroundColor: "rgba(63, 63, 63, 0.75)",
            cursor: "ew-resize",
        },
        link: {
            stroke: "rgba(170, 170, 170, 0.75)",
            strokeWidth: "3px",
            fillOpacity: 0,
        },
        connector: {
            backgroundColor: "rgba(199, 199, 41, 1)",
            borderRadius: "10px 10px 10px 10px",
            cursor: "grab",
        }
    }
};
