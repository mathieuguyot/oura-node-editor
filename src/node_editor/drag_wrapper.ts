import { XYPosition } from "./model";

type MouseMoveCb = (
    initialPos: XYPosition,
    finalPos: XYPosition,
    offSetPos: XYPosition,
    targetClassName: string
) => void;
type MouseUpCb = (targetClassName: string) => void;

export default class DragWrapper {
    private onMouseMoveCb: MouseMoveCb = () => null;
    private onMouseUpCb: MouseUpCb = () => null;
    private initialPos: XYPosition = { x: 0, y: 0 };
    private finalPos: XYPosition = { x: 0, y: 0 };
    private tmpPos: XYPosition = { x: 0, y: 0 };
    private getZoom = () => {
        return 1;
    };
    private targetClassName = "";

    constructor() {
        this.onMouseMove = this.onMouseMove.bind(this);
        this.onMouseUp = this.onMouseUp.bind(this);
    }

    onMouseDown(
        event: React.MouseEvent,
        initialPos: XYPosition,
        getZoom: () => number,
        onMouseMove: MouseMoveCb,
        onMouseUp: MouseUpCb
    ): void {
        event.stopPropagation();
        window.addEventListener("mousemove", this.onMouseMove);
        window.addEventListener("mouseup", this.onMouseUp);
        this.getZoom = getZoom;
        this.initialPos = initialPos;
        this.finalPos = { x: initialPos.x, y: initialPos.y };
        const zoom = getZoom();
        this.tmpPos = { x: event.pageX / zoom, y: event.pageY / zoom };
        this.onMouseMoveCb = onMouseMove;
        this.onMouseUpCb = onMouseUp;
    }

    private onMouseMove(event: MouseEvent) {
        const zoom = this.getZoom();
        const offsetPos = {
            x: event.pageX / zoom - this.tmpPos.x,
            y: event.pageY / zoom - this.tmpPos.y
        };
        this.finalPos.x += offsetPos.x;
        this.finalPos.y += offsetPos.y;
        this.tmpPos = { x: event.pageX / zoom, y: event.pageY / zoom };
        const targetClassName = (event.target as Element).className;
        if (typeof targetClassName === "string") {
            this.targetClassName = targetClassName;
        }
        this.onMouseMoveCb(
            { ...this.initialPos },
            { ...this.finalPos },
            offsetPos,
            this.targetClassName
        );
    }

    private onMouseUp() {
        window.removeEventListener("mousemove", this.onMouseMove);
        window.removeEventListener("mouseup", this.onMouseUp);
        this.onMouseUpCb(this.targetClassName);
    }
}
