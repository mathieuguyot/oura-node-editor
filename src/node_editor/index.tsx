export { default as NodeEditor } from "./node_editor";
export type { NodeModel, ConnectorModel, LinkModel } from "./model";
export { PinLayout } from "./model";
