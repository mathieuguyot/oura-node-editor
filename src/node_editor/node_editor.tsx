import React, { Component, RefObject } from "react";
import { produce, enableMapSet } from "immer";

import {
    NodeModel,
    XYPosition,
    LinkModel,
    LinkPositionModel,
    getLinkId,
    arePositionEquals
} from "./model";
import { Node } from "./node";
import PanZoom from "./pan_zoom";
import createLinkComponent from "./links";
import { KeyPressedWrapper } from "./events_wrappers";

enableMapSet();

type NodeEditorProps = {
    nodes: { [id: string]: NodeModel };
    links: LinkModel[];
    zoom: number;

    onNodeMove(id: string, offsetX: number, offsetY: number, offsetWidth: number): void;
    onCreateLink(link: LinkModel): void;
    setZoom(zoom: number): void;
};

type NodeEditorState = {
    linksPositions: { [linkId: string]: LinkPositionModel };
    isNodeBeingMoved: boolean;
    draggedLink?: LinkPositionModel;
    selectedNodesIds: Set<string>;
};

class NodeEditor extends Component<NodeEditorProps, NodeEditorState> {
    private nodesRefs: { [nodeId: string]: RefObject<Node> } = {};
    private keyPressedWrapper: KeyPressedWrapper = new KeyPressedWrapper();

    constructor(props: NodeEditorProps) {
        super(props);
        this.state = {
            isNodeBeingMoved: false,
            linksPositions: {},
            selectedNodesIds: new Set()
        };

        this.createReferences();

        this.getZoom = this.getZoom.bind(this);

        this.onNodeMoveStart = this.onNodeMoveStart.bind(this);
        this.onNodeMove = this.onNodeMove.bind(this);
        this.onNodeMoveEnd = this.onNodeMoveEnd.bind(this);

        this.onUpdatePreviewLink = this.onUpdatePreviewLink.bind(this);

        this.onCreateLink = this.onCreateLink.bind(this);
    }

    componentDidMount(): void {
        this.keyPressedWrapper.attachListeners();
        const { links } = this.props;
        const { linksPositions } = this.state;
        const newLinksPositions = produce(linksPositions, (draft) => {
            links.forEach((link) => {
                const linkId = getLinkId(link);
                const inputNodeRef = this.nodesRefs[link.inputNodeId].current;
                const outputNodeRef = this.nodesRefs[link.outputNodeId].current;
                if (inputNodeRef && outputNodeRef) {
                    const inputPinPosition = inputNodeRef.getConnectorPinPosition(
                        link.inputPinId,
                        link.inputPinType
                    );
                    const outputPinPosition = outputNodeRef.getConnectorPinPosition(
                        link.outputPinId,
                        link.outputPinType
                    );
                    if (inputPinPosition && outputPinPosition) {
                        draft[linkId] = { linkId, inputPinPosition, outputPinPosition };
                    }
                }
            });
        });
        this.setState({
            linksPositions: newLinksPositions
        });
    }

    componentDidUpdate(): void {
        const { links } = this.props;
        const { linksPositions } = this.state;
        let updatedIsNeeded = false;
        const newLinksPositions = produce(linksPositions, (draft) => {
            links.forEach((link) => {
                const linkId = getLinkId(link);
                const inputNodeRef = this.nodesRefs[link.inputNodeId].current;
                const outputNodeRef = this.nodesRefs[link.outputNodeId].current;
                if (inputNodeRef && outputNodeRef) {
                    const inputPinPosition = inputNodeRef.getConnectorPinPosition(
                        link.inputPinId,
                        link.inputPinType
                    );
                    const outputPinPosition = outputNodeRef.getConnectorPinPosition(
                        link.outputPinId,
                        link.outputPinType
                    );
                    if (inputPinPosition && outputPinPosition) {
                        if (
                            !(linkId in draft) ||
                            !arePositionEquals(draft[linkId].inputPinPosition, inputPinPosition) ||
                            !arePositionEquals(draft[linkId].outputPinPosition, outputPinPosition)
                        ) {
                            draft[linkId] = { linkId, inputPinPosition, outputPinPosition };
                            updatedIsNeeded = true;
                        }
                    }
                }
            });
        });
        if (updatedIsNeeded) {
            // eslint-disable-next-line react/no-did-update-set-state
            this.setState({
                linksPositions: newLinksPositions
            });
        }
    }

    componentWillUnmount(): void {
        this.keyPressedWrapper.detachListeners();
    }

    onNodeMoveStart(id: string): void {
        this.setState(
            produce((draftState: NodeEditorState) => {
                draftState.isNodeBeingMoved = true;
                if (
                    !this.keyPressedWrapper.isKeyDown("Shift") &&
                    !draftState.selectedNodesIds.has(id)
                ) {
                    draftState.selectedNodesIds.clear();
                }
                draftState.selectedNodesIds.add(id);
            })
        );
    }

    onNodeMove(offsetX: number, offsetY: number, offsetWidth: number): void {
        const { selectedNodesIds } = this.state;

        selectedNodesIds.forEach((id) => {
            const { nodes, onNodeMove } = this.props;
            const newX = nodes[id].x + offsetX;
            const newY = nodes[id].y + offsetY;
            const newWidth = nodes[id].width + offsetWidth;
            onNodeMove(id, newX, newY, newWidth > 100 ? newWidth : 100);
        });
    }

    onNodeMoveEnd(id: string, wasNodeMoved: boolean): void {
        this.setState(
            produce((draftState: NodeEditorState) => {
                draftState.isNodeBeingMoved = false;
                if (!wasNodeMoved && !this.keyPressedWrapper.isKeyDown("Shift")) {
                    draftState.selectedNodesIds.clear();
                    draftState.selectedNodesIds.add(id);
                }
            })
        );
    }

    onWheelZoom(event: React.WheelEvent): void {
        const { zoom, setZoom } = this.props;
        const { isNodeBeingMoved, draggedLink } = this.state;
        if (!isNodeBeingMoved && !draggedLink) {
            const newZoom = event.deltaY > 0 ? zoom / 1.1 : zoom * 1.1;
            setZoom(newZoom);
        }
    }

    onUpdatePreviewLink(
        inputPinPosition: XYPosition | null,
        outputPinPosition: XYPosition | null
    ): void {
        if (inputPinPosition === null || outputPinPosition === null) {
            this.setState({
                draggedLink: undefined
            });
        } else {
            this.setState({
                draggedLink: {
                    linkId: "preview",
                    inputPinPosition,
                    outputPinPosition
                }
            });
        }
    }

    onCreateLink(link: LinkModel): void {
        const { onCreateLink } = this.props;
        onCreateLink(link);
    }

    getZoom(): number {
        const { zoom } = this.props;
        return zoom;
    }

    createReferences(): void {
        const { nodes } = this.props;
        this.nodesRefs = {};
        Object.keys(nodes).forEach((key) => {
            this.nodesRefs[key] = React.createRef<Node>();
        });
    }

    render(): JSX.Element {
        const { nodes, links, zoom } = this.props;
        const { draggedLink, selectedNodesIds, linksPositions } = this.state;
        let svgDraggedLink;
        if (draggedLink) {
            svgDraggedLink = createLinkComponent({ linkType: "bezier", linkPosition: draggedLink });
        }

        const svgLinks: JSX.Element[] = [];
        links.forEach((link) => {
            const linkId = getLinkId(link);
            const linkPosition = linksPositions[linkId];
            if (linkPosition) {
                svgLinks.push(
                    createLinkComponent({
                        linkType: link.linkType,
                        key: linkId,
                        linkPosition
                    })
                );
            }
        });
        const grid = String(300 * zoom);
        return (
            <div
                onWheel={this.onWheelZoom.bind(this)}
                style={{
                    position: "relative",
                    top: "0",
                    left: "0",
                    width: "100%",
                    height: "100%",
                    overflow: "hidden",
                    backgroundColor: "#232323",
                    backgroundImage: `url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='${grid}' height='${grid}' viewBox='0 0 100 100'%3E%3Cg fill-rule='evenodd'%3E%3Cg fill='black' fill-opacity='0.4'%3E%3Cpath opacity='.5' d='M96 95h4v1h-4v4h-1v-4h-9v4h-1v-4h-9v4h-1v-4h-9v4h-1v-4h-9v4h-1v-4h-9v4h-1v-4h-9v4h-1v-4h-9v4h-1v-4h-9v4h-1v-4H0v-1h15v-9H0v-1h15v-9H0v-1h15v-9H0v-1h15v-9H0v-1h15v-9H0v-1h15v-9H0v-1h15v-9H0v-1h15v-9H0v-1h15V0h1v15h9V0h1v15h9V0h1v15h9V0h1v15h9V0h1v15h9V0h1v15h9V0h1v15h9V0h1v15h9V0h1v15h4v1h-4v9h4v1h-4v9h4v1h-4v9h4v1h-4v9h4v1h-4v9h4v1h-4v9h4v1h-4v9h4v1h-4v9zm-1 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-9-10h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm9-10v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-9-10h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm9-10v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-9-10h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm9-10v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-10 0v-9h-9v9h9zm-9-10h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9zm10 0h9v-9h-9v9z'/%3E%3Cpath d='M6 5V0H5v5H0v1h5v94h1V6h94V5H6z'/%3E%3C/g%3E%3C/g%3E%3C/svg%3E")`
                }}>
                <PanZoom getZoom={this.getZoom}>
                    <svg
                        style={{
                            position: "absolute",
                            top: "-1",
                            left: "-1",
                            width: "1",
                            height: "1",
                            overflow: "visible"
                        }}>
                        {svgLinks}
                        {svgDraggedLink}
                    </svg>
                    {Object.keys(nodes).map((key) => (
                        <Node
                            key={key}
                            ref={this.nodesRefs[key]}
                            node={nodes[key]}
                            isNodeSelected={selectedNodesIds.has(nodes[key].id)}
                            getZoom={this.getZoom}
                            onNodeMoveStart={this.onNodeMoveStart}
                            onNodeMove={this.onNodeMove}
                            onNodeMoveEnd={this.onNodeMoveEnd}
                            onCreateLink={this.onCreateLink}
                            onUpdatePreviewLink={this.onUpdatePreviewLink}
                        />
                    ))}
                </PanZoom>
            </div>
        );
    }
}

export default NodeEditor;
