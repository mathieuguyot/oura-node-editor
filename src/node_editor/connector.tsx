import React, { Component } from 'react';
import { ConnectorModel, ConnectorType, XYPosition } from './model';
import * as _ from 'lodash'
import { createConnectorComponent } from "./connector_content";

type ConnectorProps = {
    nodeId: number,
    nodeX: number,
    nodeY: number,
    width: number,
    connectorModel: ConnectorModel,

    getZoom: () => number,

    onConnectorDragStart: (connectorModel: ConnectorModel, pinPosition: XYPosition) => void,
    onConnectorMouseMove: (mousePosition: XYPosition) => void,
    onConnectorMouseUp: () => void,
    onMouseOverConnector: (connectorModel: ConnectorModel, pinPosition: XYPosition) => void,
    onMouseLeavesConnector: (mousePosition: XYPosition) => void
}

class Connector extends Component<ConnectorProps, {}>  {
    private connectorPinRef = React.createRef<HTMLDivElement>();
    private pageX: number = 0;
    private pageY: number = 0;

    constructor(props: ConnectorProps) {
        super(props);

        this.onMouseDown = this.onMouseDown.bind(this);
        this.onMouseMove = this.onMouseMove.bind(this);
        this.onMouseUp = this.onMouseUp.bind(this);
        this.onMouseOverConnector = this.onMouseOverConnector.bind(this);
        this.onMouseLeavesConnector = this.onMouseLeavesConnector.bind(this);
    }

    shouldComponentUpdate(nextProps: ConnectorProps) {
        return !_.isEqual(this.props, nextProps);
    }

    getConnectorPinPosition() : XYPosition | null {
        const pin = this.connectorPinRef.current;
        const parent = pin?.parentElement;
        if(pin && parent) {
            const widthOffset = this.props.connectorModel.connectorType === ConnectorType.Output ? this.props.width : 0;
            const x = this.props.nodeX + widthOffset;
            const y = this.props.nodeY + parent.offsetTop + (parent.offsetHeight / 2);
            return {x: x, y: y};
        }
        return null;
    }

    onMouseDown(event : React.MouseEvent) {
        event.stopPropagation();
        const pin = this.connectorPinRef.current;
        const pinPosition = this.getConnectorPinPosition();
        const zoom = this.props.getZoom();
        if(pinPosition && pin) {
            window.addEventListener("mousemove", this.onMouseMove);
            window.addEventListener("mouseup", this.onMouseUp);
            this.pageX = event.pageX / zoom;
            this.pageY = event.pageY / zoom;
            this.props.onConnectorDragStart(this.props.connectorModel, pinPosition);
        }
    }

    onMouseMove(event: MouseEvent) {
        const pin = this.connectorPinRef.current;
        const zoom = this.props.getZoom();
        if(pin) {
            let deltaX = event.pageX / zoom - this.pageX;
            let deltaY = event.pageY / zoom - this.pageY;
            this.pageX = event.pageX / zoom;
            this.pageY = event.pageY / zoom;
            this.props.onConnectorMouseMove({x: deltaX, y: deltaY});
        }
    }

    onMouseUp() {
        window.removeEventListener("mousemove", this.onMouseMove);
        window.removeEventListener("mouseup", this.onMouseUp);
        this.props.onConnectorMouseUp();
    }

    onMouseOverConnector() {
        const pinPosition = this.getConnectorPinPosition();
        if(pinPosition) {
            this.props.onMouseOverConnector(this.props.connectorModel, pinPosition);
        }
    }

    onMouseLeavesConnector(event: React.MouseEvent) {
        const mousePosition = {x: event.pageX, y: event.pageY};
        this.props.onMouseLeavesConnector(mousePosition);
    }

    render() {
        const pinPxRadius = 7;
        const pinLeftPos  = this.props.connectorModel.connectorType === ConnectorType.Input ? -pinPxRadius : this.props.width - pinPxRadius;
        return (
            <div style={{
                color: "white",
                backgroundColor: "rgba(63, 63, 63, 0.75)",
                userSelect: "none",
                MozUserSelect: "none",
                position: "relative",
                paddingLeft: pinPxRadius * 2,
                paddingRight: pinPxRadius * 2
                }}>
                    <div style={{
                        width: pinPxRadius * 2,
                        height: pinPxRadius * 2,
                        left: pinLeftPos,
                        top: "calc(50% - " + pinPxRadius + "px)",
                        backgroundColor: "rgba(199, 199, 41, 1)",
                        position: "absolute",
                        borderRadius:"10px 10px 10px 10px",
                        cursor: "grab",
                        }}
                        ref={this.connectorPinRef}
                        onMouseDown={this.onMouseDown}
                        onMouseOver={this.onMouseOverConnector}
                        onMouseLeave={this.onMouseLeavesConnector}
                    />
                    <div style={{overflow: "hidden"}}>
                        {createConnectorComponent(this.props)}
                    </div>
            </div>
        );
    }

}

export default Connector;