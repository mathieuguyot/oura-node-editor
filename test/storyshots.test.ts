// import initStoryshot from "@storybook/addon-storyshots";
import { arePositionEquals } from "../src/node_editor/model";

// initStoryshot();
test("dumb_test", () => {
    expect(arePositionEquals({ x: 1, y: 2 }, { x: 1, y: 2 })).toBe(true);
});
