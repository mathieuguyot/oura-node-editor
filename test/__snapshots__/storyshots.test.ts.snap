// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Test/Node Empty Node 1`] = `
<div
  style={
    Object {
      "backgroundColor": "red",
      "left": 0,
      "position": "relative",
      "top": 0,
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": 200,
      }
    }
  >
    <div
      onMouseDown={[Function]}
      style={
        Object {
          "MozUserSelect": "none",
          "backgroundColor": "rgba(48, 141, 97, 0.75)",
          "borderRadius": "10px 10px 0px 0px",
          "color": "white",
          "cursor": "grab",
          "height": "20px",
          "position": "relative",
          "width": 200,
        }
      }
    >
      <p
        style={
          Object {
            "overflow": "hidden",
            "paddingLeft": "10px",
          }
        }
      >
        default_node
      </p>
    </div>
    <div
      onMouseDown={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(63, 63, 63, 0.75)",
          "borderRadius": "0px 0px 10px 10px",
          "cursor": "ew-resize",
          "height": "10px",
          "position": "relative",
          "width": "200px",
        }
      }
    />
  </div>
</div>
`;

exports[`Storyshots Test/Node Full Node 1`] = `
<div
  style={
    Object {
      "backgroundColor": "red",
      "left": 0,
      "position": "relative",
      "top": 0,
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": 300,
      }
    }
  >
    <div
      onMouseDown={[Function]}
      style={
        Object {
          "MozUserSelect": "none",
          "backgroundColor": "rgba(48, 141, 97, 0.75)",
          "borderRadius": "10px 10px 0px 0px",
          "color": "white",
          "cursor": "grab",
          "height": "20px",
          "position": "relative",
          "width": 300,
        }
      }
    >
      <p
        style={
          Object {
            "overflow": "hidden",
            "paddingLeft": "10px",
          }
        }
      >
        full_node
      </p>
    </div>
    <div
      style={
        Object {
          "MozUserSelect": "none",
          "backgroundColor": "rgba(63, 63, 63, 0.75)",
          "color": "white",
          "paddingLeft": 14,
          "paddingRight": 14,
          "position": "relative",
          "userSelect": "none",
        }
      }
    >
      <div
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "backgroundColor": "rgba(199, 199, 41, 1)",
            "borderRadius": "10px 10px 10px 10px",
            "cursor": "grab",
            "height": 14,
            "left": -7,
            "position": "absolute",
            "top": "calc(50% - 7px)",
            "width": 14,
          }
        }
      />
      <div
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <div
          style={
            Object {
              "textAlign": "left",
            }
          }
        >
          x
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "MozUserSelect": "none",
          "backgroundColor": "rgba(63, 63, 63, 0.75)",
          "color": "white",
          "paddingLeft": 14,
          "paddingRight": 14,
          "position": "relative",
          "userSelect": "none",
        }
      }
    >
      <div
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "backgroundColor": "rgba(199, 199, 41, 1)",
            "borderRadius": "10px 10px 10px 10px",
            "cursor": "grab",
            "height": 14,
            "left": -7,
            "position": "absolute",
            "top": "calc(50% - 7px)",
            "width": 14,
          }
        }
      />
      <div
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <input
          style={
            Object {
              "backgroundColor": "#585858",
              "border": 0,
              "color": "white",
              "outline": "none",
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      style={
        Object {
          "MozUserSelect": "none",
          "backgroundColor": "rgba(63, 63, 63, 0.75)",
          "color": "white",
          "paddingLeft": 14,
          "paddingRight": 14,
          "position": "relative",
          "userSelect": "none",
        }
      }
    >
      <div
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "backgroundColor": "rgba(199, 199, 41, 1)",
            "borderRadius": "10px 10px 10px 10px",
            "cursor": "grab",
            "height": 14,
            "left": -7,
            "position": "absolute",
            "top": "calc(50% - 7px)",
            "width": 14,
          }
        }
      />
      <div
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <div
          style={
            Object {
              "textAlign": "left",
            }
          }
        >
          z
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "MozUserSelect": "none",
          "backgroundColor": "rgba(63, 63, 63, 0.75)",
          "color": "white",
          "paddingLeft": 14,
          "paddingRight": 14,
          "position": "relative",
          "userSelect": "none",
        }
      }
    >
      <div
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "backgroundColor": "rgba(199, 199, 41, 1)",
            "borderRadius": "10px 10px 10px 10px",
            "cursor": "grab",
            "height": 14,
            "left": 293,
            "position": "absolute",
            "top": "calc(50% - 7px)",
            "width": 14,
          }
        }
      />
      <div
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <div
          style={
            Object {
              "textAlign": "right",
            }
          }
        >
          sum
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "MozUserSelect": "none",
          "backgroundColor": "rgba(63, 63, 63, 0.75)",
          "color": "white",
          "paddingLeft": 14,
          "paddingRight": 14,
          "position": "relative",
          "userSelect": "none",
        }
      }
    >
      <div
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "backgroundColor": "rgba(199, 199, 41, 1)",
            "borderRadius": "10px 10px 10px 10px",
            "cursor": "grab",
            "height": 14,
            "left": 293,
            "position": "absolute",
            "top": "calc(50% - 7px)",
            "width": 14,
          }
        }
      />
      <div
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <div
          style={
            Object {
              "textAlign": "right",
            }
          }
        >
          product
        </div>
      </div>
    </div>
    <div
      onMouseDown={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(63, 63, 63, 0.75)",
          "borderRadius": "0px 0px 10px 10px",
          "cursor": "ew-resize",
          "height": "10px",
          "position": "relative",
          "width": "300px",
        }
      }
    />
  </div>
</div>
`;
